### DISCUSSION TITLE
bootloader script as a replacement for grub/systemd-boot/extlinux NixOS rollback feature

### DISCUSSION THREAD

**Discussion by AleXoundOS** [nix-community/NixOS-WSL#78] [Ideas]
First of all, thank you for making NixOS on WSL a reality! It works!

Originally, NixOS comes with a handy rollback feature which helps in case the system got broken. At the current (974a7c41acde9a47f1b50c2ae5239c576c1f5361) state it's even easier: making some systemd service enter a failed state [prevents the login](https://github.com/Trundle/NixOS-WSL/blob/974a7c41acde9a47f1b50c2ae5239c576c1f5361/syschdemd.sh#L30). Though, there is a workaround `wsl -u nixos` for exact this case with some nasty hacks following it related to namespaces, but I think the problem is clear.

The idea is to develop a not so complex NixOS module or maybe patch the grub/systemd-boot/extlinux module to allow a user select the system configuration to load instead of the most recent.
Without NixOS module, in the most primitive form: listing directories symlinks in `/nix/var/nix/profiles` and waiting for a user to type the desired one should suffice, residing in `syschdemd.sh` script.

**Comment by willbush** [comment]
What is the current way to rollback?

**Reply by nzbr** [reply]
Currently the only way to do a rollback is through `nix-env -p /nix/var/nix/profiles/system --rollback` (or `--switch-generation`)
EDIT: `--system` does not exist, it's `-p /nix/var/nix/profiles/system`

**Comment by AleXoundOS** [comment]
I think no way.

The proposed script is quite easy to write, though. Later I will share a variant of it.

**Comment by nzbr** [comment]
I like the idea, though having a script that awaits input in the startup process might cause problems. (For example, when the distro gets started by the vscode remote extension). The better way would probably be to add a something for this to `/bin`, so it can be executed with `wsl -d NixOS -e <name>`. This can then also be integrated with DistroLauncher, so that we can have something like `nixos --rollback` from the windows command prompt

**Reply by sielicki** [reply]
I think ergonomically this is the best solution and I'd like to contribute a PR for it. I've been waiting for input on what we actually want to implement.
